Resolve redirects when testing URLs for equality

hseg opened the pull request:
The URL DOIs resolve to can move around, with redirects pointing to the new location. To make the tests more robust, only fail if the URLs differ after redirections.

See also https://www.crossref.org/blog/urls-and-dois-a-complicated-relationship/ and item 10 in https://pardalotus.tech/posts/2024-10-02-falsehoods-programmers-believe-about-dois/

alexfikl committed:
makefile: simplify targets

alexfikl committed:
setup: switch to pyproject and hatchling

alexfikl committed:
ci: remove travis.yaml

alexfikl committed:
ci: add tests to ci

alexfikl committed:
style: fix flake8 issues

alexfikl committed:
tests: fix tests

hseg committed:
Resolve redirects when testing URLs for equality

The URL DOIs resolve to can move around, with redirects pointing to the
new location. To make the tests more robust, only fail if the URLs
differ after redirections.

See also
https://www.crossref.org/blog/urls-and-dois-a-complicated-relationship/

hseg commented:
I noticed this since the test case with https://aip.scitation.org/doi/10.1063/1.5081715 started failing

hseg commented:
Short-circuited the normalized equality so as to cut down on the number of attempts at resolving redirects -- otherwise, have been running into 403 errors that seem to arise from rate-limiting (since they go away with more attempts)

alejandrogallo commented:
Is this good to go?

-- 
Alejandro Gallo

hseg commented:
There's three points I'm unhappy/uncertain about that I'd like feedback on:
- this just makes the tests _more_ robust, but I'm unsure it entirely avoids the intermittent 403s I was getting during testing. User-agent and forcing https looks like it should work, from my testing
- it might be undesirable to unconditionally upgrade to https
- printing the HTTPException when it's caught might be unnecessary - there might be a pytest option for that. I was doing it to try to get more detail on the exception.
Other than that, I'm happy with this.

16 abr 2025 09:27:22 Alejandro Gallo ***@***.***>:

> 
>      [Imagen]*alejandrogallo* left a comment (papis/python-doi#3)[https://github.com/papis/python-doi/pull/3#issuecomment-2808514565]
>     
> <br> <br> Is this good to go?<br> <br> -- <br> Alejandro Gallo<br> <br>
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/papis/python-doi/pull/3#issuecomment-2808514565], or unsubscribe[https://github.com/notifications/unsubscribe-auth/AAJTW5IFEAHYLDGENOXZ3CL2ZXZ4TAVCNFSM6AAAAAB3GKAMRGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMBYGUYTINJWGU].
> You are receiving this because you authored the thread.
>

hseg commented:
OK, so I've addressed most of my concerns -- I've documented the problem with unconditionally upgrading to https, have made the test case warn when it needs to fall back to resolving redirects, and removed the `HTTPException` (further testing revealed that the information it provided is accessible by invoking `pytest` with `--showlocals`).
However, further testing does indeed show that the test remains flaky, with scitation abusing HTTP 403 when they want to respond with 429. I'd recommend adding `pytest-rerunfailures` or similar to the dev requirements and marking this particular test as flaky and needing to be rerun on `HTTPException`.

hseg commented:
Added `pytest-rerunfailures` to address the flakiness, I have no more notes.

hseg commented:
No good, the test still fails with 403. Can't think of what else they're probing that we need to send -- `curl -A Mozilla/5.0` works fine at the `https` endpoint.

hseg committed:
tests: Configure pytest to ignore docs

hseg commented:
OK, I've rebased this on top of #2 (including https://github.com/alexfikl/python-doi/pull/1 on the assumption @alexfikl will merge it), so now both can be merged with no issue.
I've dropped the `pytest-rerunfailures` commit, since ultimately what seems to be the correct advice is to just update the URLs when they change, as done in #2.

hseg commented:
Hm. I'm not fully satisfied by my implementation -- I *would* like to figure out how come I'm managing to get the page to load in the browser but `urllib` returns 403. I have discovered that `urllib` _is_ properly redirecting the URL, it's just that the final URL returns 403 (presumably because Cloudflare is detecting the scraper).

alexfikl commented:
> I have discovered that `urllib` _is_ properly redirecting the URL, it's just that the final URL returns 403 (presumably because Cloudflare is detecting the scraper).

Yeah, Cloudflare anti-scraper stuff is making a lot of the downloaders in Papis also fail (even though they would work properly if fed the actual webpage). I'm not sure there's much we can do there.. from what I recall the "fix" for that is to use something like [playwright](https://playwright.dev/) that can fake being a known browser.

hseg commented:
24 jun 2025 22:00:24 Alex Fikl ***@***.***>:

> 
>      [Imagen]*alexfikl* left a comment (papis/python-doi#3)[https://github.com/papis/python-doi/pull/3#issuecomment-3001556651]
>     
> 
> I have discovered that *urllib* /is/ properly redirecting the URL, it's just that the final URL returns 403 (presumably because Cloudflare is detecting the scraper).
> 
> Yeah, Cloudflare anti-scraper stuff is making a lot of the downloaders in Papis also fail (even though they would work properly if fed the actual webpage). I'm not sure there's much we can do there.. from what I recall the "fix" for that is to use something like playwright[https://playwright.dev/] that can fake being a known browser.
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/papis/python-doi/pull/3#issuecomment-3001556651], or unsubscribe[https://github.com/notifications/unsubscribe-auth/AAJTW5INPN24E7M46FNZM333FGN4PAVCNFSM6AAAAAB3GKAMRGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTAMBRGU2TMNRVGE].
> You are receiving this because you authored the thread.
> [Imagen de rastreo][https://github.com/notifications/beacon/AAJTW5OF4MY45VOFEFE2TVT3FGN4PA5CNFSM6AAAAAB3GKAMRGWGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTVS5APKW.gif]
> 

Alternatively, we could take a page out of fanficfare's book and use cloudscraper.

hseg commented:
OK, `cloudscraper` indeed seems to work:
```python
import cloudscraper
scraper = cloudscraper.create_scraper()
print(scraper.get('http://aip.scitation.org/doi/10.1063/1.5081715').url)
# https://pubs.aip.org/aip/jcp/article-abstract/150/7/074102/197572/Exact-two-component-equation-of-motion-coupled?redirectedFrom=fulltext
```

hseg committed:
Parametrize tests

This eg makes it easier to spot which particular iteration breaks

hseg committed:
Use cloudscraper to solve cloudflare challenges

Also put in a fallback using requests, but it is hacky and only works
sometimes. cloudscraper stands a better chance of consistently being
able to get to the final URL

hseg commented:
OK, added a `[challenges]` extra to bypass websites using CloudFlare (and friends, in the future) to protect against DDoS. Also added a test to check my redirect resolution code works, as @alexfikl asked for in https://github.com/papis/python-doi/pull/2#issuecomment-3001547111, though ATM it only has testcases for `cloudscraper`-dependent sites. Still, I'm happy with the code as it stands now, and would like to have it merged.
@alejandrogallo, can we get this merged?

hseg committed:
Make test_redirect cases prettier

hseg commented:
Note this still isn't perfect -- my testing must've gotten my on scitation's warnlist, because I'm starting to get redirected to their captcha page.

hseg commented:
So I tried to investigate this further today, see if I could make the test more robust.
First problem is that scitation is very aggressive with their anti-scraping, so that testcase works at best 3-4 times before their countermeasures kick in.
In trying to investigate this, I tried using different test cases to replace it.
Found some irreproducible flakiness with [this article](https://doi.org/10.1017/S0080440125100509). But even worse, trying to bulk out the testcase collection led me to discover [multiple resolution], and the papers cited there all break our current system, to a point I'm no longer sure what the correct semantics are here. Does "a canonical URL" even exist? If not, is this PR useless?

EDIT: Reading further on that page, it seems that among the multiple valid URLs for the DOI, there will always be a "primary" URL, which can obtained by adding `?locatt=mode:legacy` to the `doi.org` query string. But that means our equality needs to be at the level of DOIs, and that naively following redirects is insufficient.

[multiple resolution]: https://www.crossref.org/documentation/register-maintain-records/creating-and-managing-dois/multiple-resolution/

alejandrogallo commented:
From what I've seen this seems fine, thank you for your  work, what do you think @alexfikl ?

alexfikl commented:
@alejandrogallo Yeah, this looks good to me too! It doesn't seem like we can do much better these days + scraping seems to be disallowed more and more with AI wreaking havoc all over the place :(

It also includes #2, so I'll go ahead and close that.

>  First problem is that scitation is very aggressive with their anti-scraping, so that testcase works at best 3-4 times before their countermeasures kick in.

Might be worth looking for some other website then? The main point is to test that the DOI detection in the library works.. not necessarily that it works with `scitation` :grin:

hseg commented:
Only concern I have remaining is multiple resolution, but we could reopen that as a new issue - idk how much work it'd be to support that.
